File: leet-code/go/0704 binary-search.go
package leet

func search(nums []int, target int) int {
	return binarySearch(nums, 0, len(nums), target)
}

func binarySearch(nums []int, start, end, target int) int {
	length := end - start
	if length%2 == 0 {
		length--
	}

	pos := (length / 2) + start
	if nums[pos] == target {
		return pos
	}

	if len(nums[start:end]) == 1 {
		return -1
	}

	if nums[pos] > target {
		return binarySearch(nums, start, pos, target)
	}

	return binarySearch(nums, pos+1, end, target)
}

File: leet-code/python/0704_binary_search.py
class Solution:
  def search(self, nums: List[int], target: int) -> int:
    start, end = 0, len(nums) - 1
    while start <= end:
      i = end - start // 2
      if nums[i] == target:
        return i
      
      if nums[i] > target:
        end = i - 1
      else :
        start = i + 1
    
    return -1
    